rbd-nbd: add quiesce/unquiesce hooks

trociny opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

trociny commented:
Depends on #34243

trociny commented:
@dillaman I think it is in a good shape for review now

trociny commented:
rebased after #34243 is merged

dillaman commented on ceph.spec.in:
```diff
@@ -1289,6 +1289,7 @@ install -m 0644 -D src/logrotate.conf %{buildroot}%{_sysconfdir}/logrotate.d/cep
 chmod 0644 %{buildroot}%{_docdir}/ceph/sample.ceph.conf
 install -m 0644 -D COPYING %{buildroot}%{_docdir}/ceph/COPYING
 install -m 0644 -D etc/sysctl/90-ceph-osd.conf %{buildroot}%{_sysctldir}/90-ceph-osd.conf
+install -m 0644 -D src/tools/rbd_nbd/rbd-nbd_quiesce %{buildroot}%{_sysconfdir}/ceph/rbd-nbd_quiesce
```
Wrong permissions for a script. Also probably should live in "/usr/bin" like other helper scripts (i.e. ceph-rbdnamer). I guess this was probably so the user could tweak it, but since `rbd-nbd` now supports the `--queue-hook` optional, the end-user can create their own script anywhere.

trociny replied:
I suppose libexec directory ("/usr/libexec/rbd-nbd") would fit better for this type of scripts than "/usr/bin". Still I am not entirely sure it is better than "/etc/ceph". Yes, the reason is that the user (or rather administrator) could tweak it. Although `--quiesce-hook` is useful, the problem with this that the administrator can not enforce users to use it.

Consider a case, the administrator wants by some reason to disable quiesce (e.g. fsfreeze gets stuck frequently). With the current setup, she would just add `exit 0` as the first line to the script, so if a user run `rbd-nbd map --quiesce $image` the hook would be nop. If a user explicitly specifies `--quiesce-hook` than she is sure she wants some specific, non-default for this site, behavior. I think I would have liked to have the script in the libexec dir, instead of the  sysconfdir, if the rbd-nbd had had a config, where the administrator could specify the hook location. But we don't have such. In other words I would like the administrator has a possibility to configure the site default behavior for "quiesce", and it seems the only way to do this is to have the default `--quiesce-hook` location to be in the sysconfig ("/etc") dir.

It think having an executable script in '/etc' is rather common practice. Usually you can find many scripts there (init, network config, ...), and we already have rbdmap, which I actually used as an example for `rbd-nbd_quiesce`.

I think I would be fine to have rbd-nbd_quiesce in  the libexec dir and have "/etc/ceph/rbd-nbd_quiesce" as a symlink to it. Not sure though we really want this complication.

dillaman replied:
All the NBD commands require `CAP_SYS_ADMIN` if I remember correctly, so it would not be a random user running `rbd-nbd`. You could add a config option for a default `quiesce-hook` location so that it can be overwritten globally, per-host, etc. 

Looks like we install `rbdmap` in `/usr/bin`. The new systemd scripts point to `/usr/bin`. The `rbdmap` that goes in `/etc/ceph` is the config file.

trociny replied:
A "user" could be just a third party software used to map rbd-nbd devices, which calls (may call) rbd-nbd map with `--quiesce` option but does not provide a possibility to pass `--quiesce-hook` option. 

> Looks like we install rbdmap in /usr/bin. The new systemd scripts point to /usr/bin. The rbdmap that goes in /etc/ceph is the config file.

Indeed. I just had a wrong impression about it then.

So you still prefer having it somewhere in bin or libexec? I am not going to insist. It is a little sad to me that it would not be possible to configure a system in such a way that a user could just call `rbd-nbd map --quiesce` and should always add `--quiesce-hook` if they want to use their own hook. But I can live with this -- anyway most likely user will not call the command directly but use some third party tool. (And having the hook in /etc has its own drawbacks: e.g. to make sure it is not overwritten on the upgrade).

dillaman replied:
> So you still prefer having it somewhere in bin or libexec?

Yes, that is what I would lean towards. If you then add a config option for the default hook location, it seems like you could then configure a node to always use a user-provided hook just as easily as if the user tweaked the package-provided hook.

trociny replied:
Ok. I will move the hook to libexec. Not sure about the config option though (I suppose you mean ceph.conf, not a separate nbd config), because before this we have not have any config options for nbd and I am not sure we really want (mostly the same arguments as for using internal librbd functions).

dillaman replied:
Yeah, I was talking about a `common/config.cc` addition to support MGR-based global config overrides + `ceph.conf` + CLI processing. I am not really pushing for it, but it offers a way to have a default hook override location and natively handles a hook override w/o the need to tweak `rbd device map`.

trociny replied:
I have changed the default hook location to be in  the rbd-nbd libexec directory. 

I have not added the config option though. It is not that I am against it, I am just reluctant to open this can of worms right now: if we add an rbd-nbd option for this, what about other rbd-nbd options? It would be logical to add them too. And if we add nbd options, what about kernel options, and ggate options? It would be logical to add them too...

So it can be done later together with other options or may be we think out some other solution.

dillaman replied:
:+1:

trociny commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -380,6 +393,66 @@ class NBDServer
+
+    dout(20) << __func__ << dendl;
+
+    int r = image.flush();
```
@dillaman There is an issue with this `flush` I forgot to mention. I think it may abuse `rbd_cache_writethrough_until_flush`. Do you think it is the problem we need to solve? Do you see a way?

dillaman replied:
Probably should avoid the flush if you haven't seen a flush  -- or perhaps hide it behind the librbd API (i.e. after you get the `quiesce_complete` callback, kick an internal flush).

trociny replied:
It seems currently the rbd-nbd uses only librbd API functions and I would not want to break this until it is really necessary.

dillaman replied:
I'm probably not understanding. I'm saying you probably should  (1) not issue the `flush` unless you are tracking if `rbd-nbd` has already seen a flush or (2) or just issue an internal flush within librbd's `quiesce_complete` call.

trociny replied:
I mean, how can I tell from rbd-nbd, just using librbd API, if `flush` is already issued or issue "internal" flush? It seems I need to use internal functions for this (which I really would not want) or extend the API.

And I am not sure about option (1) -- what I want here is to make sure all my pending io is complete, the `flush` call just looked like the easiest way. Not issuing `flush` I just would not know if (when) it is complete.

dillaman replied:
You'd have to have `rbd-nbd` watch for an incoming flush and just flip a bit to know that flushes are safe. The nice thing about the internal flush is that it won't enable writeback since it's flagged as an internal-source (and therefore no different than a `block_writes` call that already happens upon snapshot create).

trociny replied:
> You'd have to have rbd-nbd watch for an incoming flush and just flip a bit to know that flushes are safe. 

Indeed. Though I am still not very happy with this solution, as it does not look like exactly what I want here.

> The nice thing about the internal flush is that it won't enable writeback since it's flagged as an internal-source (and therefore no different than a block_writes call that already happens upon snapshot create).

Yes, this is what I want but via API. I.e. something like `int rbd_flush2(bool dont_abuse_writethrough_until_flush)` but adding it just for this purposes looks like overkill.

I think I will do the option (1) (which is not ideal but easy to implement).

dillaman replied:
Just for my curiosity, what's your objection to option 2?

trociny replied:
For the option 2 (issue internal flash) I would need to use librbd internal functions, which I reluctant to do, because currently rbd-nbd uses only API functions. Well, it uses many things from common and global that makes it impossible to ship it independently (I think), so using an internal function would not change much. But there are other points why it is nice to keep it to use "pure" librbd API: it may serve as an example for other third party applications how to use librbd and actually we know ourselves what librbd functionality is useful and what is missing. So for the option (2) we miss API to call the flush as if it is flagged as an internal-source, i.e. without enabling the writeback cache. We could add this functionality but I am not sure it is worth doing for this case. But if you think it is useful I will add `flush2` or something like this.

dillaman replied:
Sorry, I was saying tweak the implementation of the `quiesce_complete` API call to (internally) issue an internal flush.

trociny replied:
Ah, interesting. Though I am not sure I like this much. I am not sure it would be obvious for users what `quiesce_complete` does and for. And I am not sure users would always want an internal flush is issued there. 

(Actually, I think even for rbd-nbd this flush is not necessary if the hook works properly, because I expect the hook (fsfreeze) would call it and wait for complete. It is here just as a safety guard, just in case the hook misbehaves.) 

And adding `sync` to `quiesce_complete` would mean `quiesce_complete` might block and I suppose we would need async version of it too (this is not need for rbd-nbd right now because it has a separate thread for it, but it might change).

dillaman replied:
Fair enough.

trociny replied:
Ok. I added a commit that skips flush if it is unsafe to call it. If you still prefer a different solution please let me know.

trociny commented:
@dillaman updated. I think I addressed all your comments but that one about quiesce hook location. If after my arguments you still not happy with /etc/ceph I will update this too.

Also, after additional testing, I updated `rbd-nbd_quiesce` script: it did not work for ubuntu because `findmnt` and `fsfreeze` had different location. I fixed it by using PATH to look for the tools.

And I improved a little the qa script: the test tries to detect if the default quiesce hook is executed without errors by parsing the rbd-nbd log.

trociny commented:
@dillaman added a commit that enables random distros for nbd teuthology test, so it is tested not only for ububtu.

The results: http://pulpito.ceph.com/trociny-2020-05-08_09:57:17-rbd-wip-mgolub-testing-distro-basic-smithi/

dillaman commented on src/tools/rbd_nbd/rbd-nbd_quiesce:
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
```
... any thoughts on this?

dillaman commented:
Looks like `ceph_test_librbd` failed under `TestLibRBD.QuiesceWatchTimeout` 

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-05-14_16:53:51-rbd-wip-jd-testing-distro-basic-smithi/5055019/teuthology.log

trociny commented:
> Looks like ceph_test_librbd failed under TestLibRBD.QuiesceWatchTimeout

The problem was with the test. On teuthology 13 sec timeout was not always enough for     "rbd_quiesce_notification_attempts = 2": if a socket failure is injected after quiesce notification is sent the request is resent and the notifier timeout (and as a result the next notification attempt) is delayed. I reproduced this with debug on, see the extract from the log below.

I added a patch here that improves the test.

```
2020-05-15T11:24:55.579+0000 7f00a0fe2b00  5 librbd::Operations: 0x563b673ec490 snap_create: snap_name=snap2
2020-05-15T11:24:55.579+0000 7f00a0fe2b00 20 librbd::Operations: send_local_request
2020-05-15T11:24:55.579+0000 7f00a0fe2b00  5 librbd::Operations: 0x563b673ec490 execute_snap_create: snap_name=snap2
2020-05-15T11:24:55.579+0000 7f00a0fe2b00  5 librbd::SnapshotCreateRequest: 0x7f0028028c30 send_notify_quiesce
2020-05-15T11:24:55.579+0000 7f00a0fe2b00 10 librbd::ImageWatcher: 0x7f0024002f20 notify_quiesce: request_id=4
2020-05-15T11:24:55.579+0000 7f00a0fe2b00 10 librbd::ImageWatcher: 0x7f0024002f20 notify_quiesce: async_request_id=[4721,139639654453328,4] attempts=2
2020-05-15T11:24:55.579+0000 7f00a0fe2b00 20 librbd::watcher::Notifier: 0x7f0024002fa0 notify: pending=1
2020-05-15T11:24:55.580+0000 7f003f7fe700  5 librbd::Watcher: 0x7f0024002f20 notifications_blocked: blocked=0
2020-05-15T11:24:55.580+0000 7f003f7fe700 10 librbd::Watcher::C_NotifyAck 0x7f002400b0a0 C_NotifyAck: id=1756641624146, handle=139639654453328
2020-05-15T11:24:55.580+0000 7f003f7fe700 10 librbd::ImageWatcher: 0x7f0024002f20 quiesce request: [4721,139639654453328,4]
2020-05-15T11:24:55.580+0000 7f003f7fe700 20 librbd::ImageState: 0x7f006400eaf0 notify_quiesce
2020-05-15T11:24:55.580+0000 7f003f7fe700 20 librbd::ImageState: 0x7f002c051f00 QuiesceWatchers::notify event: 0
2020-05-15T11:24:55.580+0000 7f003d7fa700 20 librbd::ImageState: 0x7f002c051f00 QuiesceWatchers::operator(): handle=0, event_type=0
2020-05-15T11:24:59.556+0000 7f00557fa700  0 -- 172.21.15.182:0/2384116019 >> [v2:172.21.15.124:6826/27120,v1:172.21.15.124:6827/27120] conn(0x7f002400a720 msgr2=0x7f002400da10 crc :-1 s=STATE_CONNECTION_ESTABLISHED l=1).read_until injecting socket failure
2020-05-15T11:24:59.558+0000 7f003f7fe700  5 librbd::Watcher: 0x7f0024002f20 notifications_blocked: blocked=0
2020-05-15T11:24:59.558+0000 7f003f7fe700 10 librbd::Watcher::C_NotifyAck 0x7f0024015610 C_NotifyAck: id=1756641624146, handle=139639654453328
2020-05-15T11:24:59.558+0000 7f003f7fe700 10 librbd::ImageWatcher: 0x7f0024002f20 duplicate quiesce request: [4721,139639654453328,4]
2020-05-15T11:24:59.558+0000 7f003f7fe700  5 librbd::Watcher: 0x7f0024002f20 notifications_blocked: blocked=0
2020-05-15T11:24:59.559+0000 7f003f7fe700 10 librbd::Watcher::C_NotifyAck 0x7f002400b0a0 C_NotifyAck: id=1760936591443, handle=139639654453328
2020-05-15T11:24:59.559+0000 7f003f7fe700 10 librbd::ImageWatcher: 0x7f0024002f20 duplicate quiesce request: [4721,139639654453328,4]
2020-05-15T11:25:04.558+0000 7f003f7fe700 20 librbd::watcher::Notifier: 0x7f0024002fa0 handle_notify: r=-110
2020-05-15T11:25:04.559+0000 7f003f7fe700 20 librbd::watcher::Notifier: 0x7f0024002fa0 handle_notify: pending=0
2020-05-15T11:25:04.559+0000 7f003d7fa700 10 librbd::ImageWatcher: 0x7f0024002f20 operator(): async_request_id=[4721,139639654453328,4] timed out
2020-05-15T11:25:04.559+0000 7f003d7fa700 10 librbd::ImageWatcher: 0x7f0024002f20 notify_quiesce: async_request_id=[4721,139639654453328,4] attempts=1
```

trociny commented:
Actually, I decided to push the test fix as a separate PR (#35089), as it is not much related to this PR.

dillaman commented:
> Actually, I decided to push the test fix as a separate PR (#35089), as it is not much related to this PR.

Ack -- thanks.
